Zend/zend_alloc.c

```diff
@@ -1900,10 +2027,108 @@ static void zend_mm_free_huge(zend_mm_heap *heap, void *ptr ZEND_FILE_LINE_DC ZE
 #endif
 }
 
+/********/
+/* Rand */
+/********/
+
+/* Xoshiro256** PRNG based on implementation from Go Kudo in
+ * ext/random/engine_xoshiro256starstar.c, based on code from David Blackman,
+ * Sebastiano Vigna. */
+
+static inline uint64_t splitmix64(uint64_t *seed)
+{
+	uint64_t r;
+
+	r = (*seed += 0x9e3779b97f4a7c15ULL);
+	r = (r ^ (r >> 30)) * 0xbf58476d1ce4e5b9ULL;
+	r = (r ^ (r >> 27)) * 0x94d049bb133111ebULL;
+	return (r ^ (r >> 31));
+}
+
+ZEND_ATTRIBUTE_CONST static inline uint64_t rotl(const uint64_t x, int k)
+{
+	return (x << k) | (x >> (64 - k));
+}
+
+static inline uint64_t zend_mm_rand_generate(zend_mm_rand_state *s)
+{
+	const uint64_t r = rotl(s->state[1] * 5, 7) * 9;
+	const uint64_t t = s->state[1] << 17;
+
+	s->state[2] ^= s->state[0];
+	s->state[3] ^= s->state[1];
+	s->state[1] ^= s->state[2];
+	s->state[0] ^= s->state[3];
+
+	s->state[2] ^= t;
+
+	s->state[3] = rotl(s->state[3], 45);
+
+	return r;
+}
+
+static inline void zend_mm_rand_seed256(zend_mm_rand_state *state, uint64_t s0, uint64_t s1, uint64_t s2, uint64_t s3)
+{
+	state->state[0] = s0;
+	state->state[1] = s1;
+	state->state[2] = s2;
+	state->state[3] = s3;
+}
+
+static inline void zend_mm_rand_seed64(zend_mm_rand_state *state, uint64_t seed)
+{
+	uint64_t s[4];
+
+	s[0] = splitmix64(&seed);
+	s[1] = splitmix64(&seed);
+	s[2] = splitmix64(&seed);
+	s[3] = splitmix64(&seed);
+
+	zend_mm_rand_seed256(state, s[0], s[1], s[2], s[3]);
+}
+
 /******************/
 /* Initialization */
 /******************/
 
+static zend_result zend_mm_refresh_key(zend_mm_heap *heap)
+{
+	heap->shadow_key = (uintptr_t) zend_mm_rand_generate(&heap->rand_state);
+
+	return SUCCESS;
+}
+
+static zend_result zend_mm_init_key(zend_mm_heap *heap)
+{
+	uint64_t seed[4];
+	char errstr[128];
+	if (zend_os_csprng_random_bytes(&seed, sizeof(seed), errstr, sizeof(errstr)) == SUCCESS) {
+		zend_mm_rand_seed256(&heap->rand_state,
+				seed[0], seed[1], seed[2], seed[3]);
+	} else {
```

Comment:
One note of caution: xoshiro256** fails catastrophically if the state is all-zeroes, which happens with if a chance of 2^(-256). It can likely be ignored here, but is explicitly handled in ext/random for full correctness.

Leaving the decision up to you what to do with it, just wanted to note it for completeness.